### PR TITLE
Pyi 685

### PR DESCRIPTION
**Pull request by SunilDGDS**
### What changed
In access token handler, a jwt claimset is received as part of the request body . This claimset is read and validated against configuration values coming from parameter store .

- [PYI-685](https://govukverify.atlassian.net/browse/PYI-685)

### Other considerations

The clientid from the claimset is used to read the configuration .  It is possible that the client does not need to be authenticated ('ATHENTICATION_METHOD set to none in the config) , therefore not receiving the claimset as part of the response body in this case . How do we know that can we read the configuration in advance to check if the client's authentification method ?

### CODE REVIEW

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java [hunk 1]
```diff
@@ -204,22 +205,21 @@ void shouldReturn400WhenInvalidJwtProvided() throws Exception {
         APIGatewayProxyRequestEvent event = new APIGatewayProxyRequestEvent();
         event.setBody(tokenRequestBody);
 
-        ValidationResult validationResultError =
-                new ValidationResult<>(false, OAuth2Error.UNSUPPORTED_GRANT_TYPE);
+        ValidationResult<ErrorResponse> validationResultError =
+                new ValidationResult<>(false, ErrorResponse.INVALID_REQUEST_PARAM);
```
Comment (Wynndow) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java]: I think this can be dropped - doesn't seem to be used.

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java [hunk 2]
```diff
@@ -230,22 +230,21 @@ void shouldReturn400WhenJwtMissingFromRequestProvided() throws Exception {
         APIGatewayProxyRequestEvent event = new APIGatewayProxyRequestEvent();
         event.setBody(tokenRequestBody);
 
-        ValidationResult validationResultError =
-                new ValidationResult<>(false, OAuth2Error.INVALID_REQUEST_OBJECT);
+        ValidationResult<ErrorResponse> validationResultError =
+                new ValidationResult<>(false, ErrorResponse.INVALID_REQUEST_PARAM);
```
Comment (Wynndow) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java]: Same here, I don't think it's used.

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java [hunk 3]
```diff
@@ -1,37 +1,118 @@
...
+
+    private String getClientAuthenticationMethod(String clientId)
+            throws ClientAuthenticationException {
+        String clientAuthenticationMethod =
+                configurationService.getClientAuthenticationMethod(clientId);
+        if (clientAuthenticationMethod == null) {
+            LOGGER.error("Failed to find config for client ID: {}", clientId);
+            throw new ClientAuthenticationException(
+                    String.format("Config for client ID '%s' not found", clientId));
         }
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: The ssmProvider in the config service will throw a runtime error if it can't find a value in the parameter store, so i think this block can be safely dropped.

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java [hunk 4]
```diff
@@ -1,37 +1,118 @@
...
-        return ValidationResult.createValidResult();
+        return clientAuthenticationMethod;
+    }
+
+    private void validateExpiration(JWTAuthenticationClaimsSet claimsSet)
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: Could this be called something else to make it obvious that it's validating the expiry isn't too far in the future?

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java [hunk 5]
```diff
@@ -1,37 +1,118 @@
...
+        OffsetDateTime offsetDateTime =
+                OffsetDateTime.now().plusSeconds(Long.parseLong(maxAllowedTtl));
+        if (expirationTime.getTime() / 1000L > offsetDateTime.toEpochSecond()) {
+            LOGGER.error("Client JWT expiry date is too far in the future");
+            throw new InvalidClientException(
+                    "The client JWT expiry date has surpassed the maximum allowed ttl value");
+        }
+    }
+
+    public static ClientAuthenticationVerifier<Object> getClientAuthVerifier(
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: As this is being called in the constructor I don't think it needs to be static. I think it was originally intended to be called externally and the return value passed into the constructor to allow injection. As we're not doing that I reckon just make it an instant method.

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/ConfigurationService.java [hunk 6]
```diff
@@ -184,6 +189,13 @@ public Certificate getClientCertificate(String clientId) throws CertificateExcep
                         System.getenv("ENVIRONMENT"), clientId));
     }
 
+    public String getClientAuthenticationMethod(String clientId) {
+        return getParameterFromStore(
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/service/ConfigurationService.java]: Should we just use `ssmProvider.get()` rather than this wrapper method?

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java [hunk 7]
```diff
@@ -1,37 +1,118 @@
...
+                    throw new ClientAuthenticationException(
+                            String.format(
+                                    "Missing client_assertion jwt for configured client '%s'",
+                                    clientId));
+                }
             }
-        } catch (ParseException e) {
-            LOGGER.error("Unable to parse Claims set {} ", e.getMessage());
-            return new ValidationResult<>(false, OAuth2Error.INVALID_CLIENT);
+        } else {
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: This is quite hard to follow through all the branches - could it be refactored to a well name private method?

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java (hunk 1 above)
Comment (dannyhayes-gds) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java]: Yeah good catch 👍

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java (hunk 4 above)
Comment (dannyhayes-gds) [reply lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: Yeah sounds good, I'll change this.

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/ConfigurationService.java (hunk 6 above)
Comment (dannyhayes-gds) [reply lib/src/main/java/uk/gov/di/ipv/core/library/service/ConfigurationService.java]: We can do? I think I mostly just followed the pattern of some of the other methods we have doing it like this. But I agree its not really adding much with the wrapper method

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java (hunk 7 above)
Comment (dannyhayes-gds) [reply lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: Yeah good shout, I'll refactor this slightly

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java (hunk 1 above)
Comment (SunilDGDS) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java]: Removed

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java (hunk 2 above)
Comment (SunilDGDS) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java]: Removed

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java [hunk 8]
```diff
@@ -40,59 +40,59 @@ public TokenRequestValidator(ConfigurationService configurationService) {
     public void authenticateClient(String requestBody, Map<String, String> queryParams)
             throws ClientAuthenticationException {
         if (!queryParams.containsKey(CLIENT_ASSERTION_PARAM)) {
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: Just to be super picky, could this be:
```
if (queryParams.containsKey(CLIENT_ASSERTION_PARAM)) {
    authenticateClientWithJwt(requestBody);
else {
    authenticateClientWithoutJwt(queryParams);
}
```

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java [hunk 9]
```diff
@@ -40,59 +40,59 @@ public TokenRequestValidator(ConfigurationService configurationService) {
...
-
-                clientId = clientJwt.getClientID().getValue();
+            authenticateClientWithJwt(requestBody);
+        }
+    }
 
-                String clientAuthenticationMethod = getClientAuthenticationMethod(clientId);
+    private void authenticateClientWithoutJwt(Map<String, String> queryParams)
+            throws ClientAuthenticationException {
+        if (!queryParams.containsKey(CLIENT_ID_PARAM)) {
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: Again, could we use `if (queryParams.containsKey(CLIENT_ID_PARAM))` and switch the blocks?

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java (hunk 3 above)
Comment (SunilDGDS) [reply lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: refactor to catch ssmProvider exception so that tests do not fall over

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java (hunk 4 above)
Comment (SunilDGDS) [reply lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: renamed

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java (hunk 5 above)
Comment (SunilDGDS) [reply lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: change to concrete

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/ConfigurationService.java (hunk 6 above)
Comment (SunilDGDS) [reply lib/src/main/java/uk/gov/di/ipv/core/library/service/ConfigurationService.java]: changed to use ssmProvider

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java (hunk 7 above)
Comment (SunilDGDS) [reply lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: Refactor into 2 distincts smaller method call

File: lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java (hunk 8 above)
Comment (dannyhayes-gds) [reply lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java]: 👍  yeah agree that would be nicer

**Review by Wynndow — APPROVED** [review]
